corehq/preindex/django_migrations.py

```diff
@@ -0,0 +1,99 @@
+import hashlib
+import os
+from operator import attrgetter
+from pathlib import Path
+
+from couchdbkit.designer import document
+
+from django.core.management import call_command
+from django.db.migrations import RunPython
+
+from .accessors import get_preindex_designs
+
+LOCK_PATH = Path(__file__).parent / "couchviews.lock"
+LOCK_PREAMBLE = """\
+# This file is autogenerated every time a new "reindex" migration is
+# generated by the `makemigrations` management command. If it becomes
+# outdated it can be updated by running:
+#
+# ./manage.py makemigrations preindex
+#
+# The commcare-hq automated tests require that the contents of this
+# file match the current Couch views list. If that check fails on a
+# GitHub PR, the PR will not be eligible for merging. You can check if
+# the current state of this file will pass automated tests by running:
+#
+# ./manage.py makemigrations preindex --check --dry-run
+#
+# NOTE: Some files listed here are not meant to be included in Couch
+# design documents (e.g., Python source files), but are included as a
+# result of a bug in corehq.couchapps.
+#
+"""
+
+
+class ReindexCouchViews(RunPython):
+    """Reindex Couch DB views and Elasticsearch indexes
+
+    Use as needed whenever Couch views are changed and need to be
+    reindexed. This operation may take a long time, depending on what is
+    being reindexed.
+
+    It is planned for the Elasticsearch index building features to move
+    to migration operations dedicated to that purpose alone.
+    See corehq/apps/es/migration_operations.py
+    """
+
+    def __init__(self):
+        super().__init__(self.run, RunPython.noop)
+
+    def deconstruct(self):
+        return (self.__class__.__qualname__, [], {})
+
+    def run(self, apps, schema_editor):
+        call_command("preindex_everything", 8)
+        call_command("sync_finish_couchdb_hq")
+        call_command("ptop_es_manage", flip_all_aliases=True)
+
+
+def iter_couch_lock_lines(preamble=LOCK_PREAMBLE):
+    yield from preamble.splitlines(keepends=True)
+    seen = set()
+    corehq = Path(__file__).parent.parent
+    assert corehq.name == "corehq", corehq
+    rootlen = len(str(corehq.parent)) + 1
+    for design in sorted(get_preindex_designs(), key=attrgetter("design_path")):
+        doc = document(design.design_path).doc()
+        for relpath in doc["couchapp"]["manifest"]:
+            fullpath = os.path.join(design.design_path, relpath)
+            viewpath = fullpath[rootlen:]
+            if os.path.isdir(fullpath):
+                continue
+            sha = sha256sum(fullpath)[:20]
+            if (sha, viewpath) not in seen:
+                yield f"{sha} {viewpath}\n"
+                seen.add((sha, viewpath))
+
+
+def designs_did_change(lines, lock_path=LOCK_PATH):
+    if not os.path.exists(lock_path):
+        return True
+    with open(lock_path, "r") as file:
+        return list(file) != lines
```

Comment:
Is the order of `get_preindex_designs()` deterministic? In other words: if the lines only change in _order_ (if that's even possible), does that justify a "the designs have changed" condition?

Reply:
Good catch!

> Is the order of `get_preindex_designs()` deterministic?

Yes. "Designs" are sorted by design path:

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/corehq/preindex/django_migrations.py#L65

However, Couch design doc manifests apparently were apparently not sorted. Fixed in https://github.com/dimagi/commcare-hq/pull/32544/commits/6d503ca991fd619653212f6053d0a736402d035b